Add velocity array support for trajectories scan

LeandroMartinsdS opened the pull request:
This PR is related to the working in progress to remove the velocities calculation from the motion program, and bring it to the driver level.

Some points that still need to be addressed, some of them discussed with @coretl , @ajgdls  and  @JimboMonkey :

- [x] Fix white spaces to make it clear where are the changes
- [x] Implement the velocities calculation in the driver level instead of in motion program
  - [x] Rollback appendVelMode/appendVelArray to append
  - [x] Substitute VelType for CalcVel, and remove from the broker
  - [x] Set CalcVel using asyn
  - [x] Port the equations from the motion program
- [x] Stop sending the VelMode
- [x] Update motion programs removing the velocities calculations (and correspondent subprograms)
- [x] Update the memory buffer
- [x] Bump motion program version
- [x] Update tests

If I'm missing something, please let me know.

LeandroMartinsdS committed:
Add velocity array support for trajectories scan

ajgdls commented:
Hi @LeandroMartinsdS I agree with your bullet list of tasks to complete this implementation.

LeandroMartinsdS committed:
Rollback to pmacTrajectory::append

LeandroMartinsdS committed:
Add method for velocity calculation

LeandroMartinsdS committed:
Update records in the trajectory template

LeandroMartinsdS committed:
Update hardware interfaces

LeandroMartinsdS committed:
Update motion programs

LeandroMartinsdS committed:
Update unitTests

ajgdls commented:
Hi @LeandroMartinsdS at the moment it is not possible to download the motion programs to the PMAC.  I think there are a few syntax errors in the pmc files.  I would recommend the following checks:
- Remove any commented lines (I notice a few where there is a "//" but not a ";") to prevent download errors
- There are still instances of using variables (e.g. Prev_A) but these are no longer defined in the script as they have been commented out.

LeandroMartinsdS committed:
Remove commented lines and undeclared variables

LeandroMartinsdS committed:
Update memory check for Clipper

LeandroMartinsdS committed:
Add missing update in velocity buffer address

LeandroMartinsdS committed:
Fix address shift in velocity buffers

LeandroMartinsdS committed:
Update Clipper BuffLen

LeandroMartinsdS committed:
Remove unnecessary tScanCalcVel_ fixing warning

LeandroMartinsdS committed:
Split the loop to fill the velocities array

For the Velocities calculation it's necessary that the positions array is already filled

LeandroMartinsdS committed:
Add times check to avoid division by zero

LeandroMartinsdS committed:
Tidy up using the defined macros

LeandroMartinsdS committed:
Set as default to execute the velocity calculations

LeandroMartinsdS committed:
Remove initial value from the Calculation Velocity flag

LeandroMartinsdS committed:
Fix axis mask shift for velocities command buffer

LeandroMartinsdS committed:
Fix missing return in trajectory motion program

LeandroMartinsdS committed:
Fix axes Vel_Adr memory in Clipper trajectory code

ajgdls commented on pmacApp/pmc/trajectory_scan_clipper.pmc:
```diff
@@ -2,7 +2,7 @@
 ; *****************************************************************************************
 #define VarAdr       B8     ; Prefix of address (eg VarAdr=300 then $30000 is the start address)
-#define BuffLen      950    ; Length of buffers
+#define BuffLen      500    ; Length of buffers
```
The amount of required memory has increased by close to a factor of two, and this buffer length number has decreased but not by factor of two.  Have you checked this fits within the allowed memory space?

LeandroMartinsdS replied:
The number of buffers had increased from 10 to 19 by adding the velocities for each axis.
The buffer length has decreased applying the ratio of 10/19, resulting in the buffer length of 500.
By my tests, the buffer fits even for a value close to 530, but supposing that already has some conservative margin I keep the value of 500.

ajgdls replied:
OK perfect.

ajgdls commented on pmacApp/pmc/trajectory_scan_code_ppmac.pmc:
```diff
@@ -137,57 +137,48 @@ N102:
-        Prev_W = Current_W
         Current_W = Next_W(CalculatedBase)
-        W_Prev_Vel = W_Vel
+        V_Vel = Next_V_Vel(CalculatedBase)
```
This needs to map W velocities

ajgdls commented on pmacApp/src/pmacController.cpp:
```diff
@@ -726,6 +738,7 @@ void pmacController::setupBrokerVariables(void) {
   pBroker_->addReadVariable(pmacMessageBroker::PMAC_SLOW_READ, PMAC_TRAJ_BUFFER_LENGTH);
   pBroker_->addReadVariable(pmacMessageBroker::PMAC_SLOW_READ, PMAC_TRAJ_BUFF_ADR_A);
   pBroker_->addReadVariable(pmacMessageBroker::PMAC_SLOW_READ, PMAC_TRAJ_BUFF_ADR_B);
+  // pBroker_->addReadVariable(pmacMessageBroker::PMAC_SLOW_READ, PMAC_TRAJ_VELOCITY_TYPE);
```
Added but then left commented out, is this required?

ajgdls commented on pmacApp/src/pmacController.cpp:
```diff
@@ -1350,7 +1368,7 @@ asynStatus pmacController::mediumUpdate(pmacCommandStore *sPtr) {
             gpioOutputs += gpioBit << gpio;
           }
-        }
+         }
```
Remove incorrect whitespace

ajgdls commented on pmacApp/src/pmacController.cpp:
```diff
@@ -2740,6 +2790,7 @@ asynStatus pmacController::buildProfile(int csNo) {
 
     // Check for any invalid times
     int maxValue = 0xFFFFFF;
+    // TODO:  Add cid check for pppmac
```
Has this been completed?

ajgdls commented on pmacApp/src/pmacController.cpp:
```diff
@@ -2781,10 +2832,10 @@ asynStatus pmacController::buildProfile(int csNo) {
     // Initialise the trajectory store
     status = pTrajectory_->initialise(numPoints);
-
+    
```
Whitespace

ajgdls commented on pmacApp/src/pmacController.cpp:
```diff
@@ -4486,6 +4625,29 @@ asynStatus pmacMonitorVariables(const char *controller, const char *variablesStr
   return asynSuccess;
 }
 
+// /**
+//  * Set as true the PMAC controller boolean
+//  * tScanUseVelArray_ to use the velocity array in trajectory scans.
+//  * Default value is false.
+//  * @param controller The Asyn port name for the PMAC controller.
+//  */
+// asynStatus pmacSetTrajectoryUseVelArray(const char *controller)
+// {
+//   pmacController *pC = NULL;
+
+//   static const char *functionName = "pmacSetTrajectoryUseVelArray";
+
+//   pC = (pmacController *)findAsynPortDriver(controller);
+//   if (!pC)
+//   {
+//     printf("%s:%s: Error port %s not found\n",
+//            driverName, functionName, controller);
+//     return asynError;
+//   }
+
+//   return pC->pmacSetTrajectoryUseVelArray();
+// }
+
```
Remove this commented code if it is not required?

ajgdls commented on pmacApp/unitTests/test_PMACTrajectory.cpp:
```diff
@@ -86,14 +96,19 @@ BOOST_AUTO_TEST_CASE(test_PMACTrajectory)
-  int velocityVal;
-  BOOST_CHECK_EQUAL(trajectory.getVelocityMode(-1, &velocityVal), asynError);
-  BOOST_CHECK_EQUAL(trajectory.getVelocityMode(20, &velocityVal), asynError);
+  // int velocityVal;
+  // BOOST_CHECK_EQUAL(trajectory.getVelocityMode(-1, &velocityVal), asynError);
+  // BOOST_CHECK_EQUAL(trajectory.getVelocityMode(20, &velocityVal), asynError);
```
Remove commented code

ajgdls commented on pmacApp/unitTests/test_PMACTrajectory.cpp:
```diff
@@ -104,18 +119,23 @@ BOOST_AUTO_TEST_CASE(test_PMACTrajectory)
-  BOOST_CHECK_EQUAL(velocityVal, 5);
-  BOOST_CHECK_EQUAL(trajectory.getVelocityMode(14, &velocityVal), asynSuccess);
-  BOOST_CHECK_EQUAL(velocityVal, 5);
+  // BOOST_CHECK_EQUAL(trajectory.getVelocityMode(4, &velocityVal), asynSuccess);
+  // BOOST_CHECK_EQUAL(velocityVal, 5);
+  // BOOST_CHECK_EQUAL(trajectory.getVelocityMode(14, &velocityVal), asynSuccess);
+  // BOOST_CHECK_EQUAL(velocityVal, 5);
```
Remove commented code

ajgdls commented on pmacApp/unitTests/test_PMACTrajectory.cpp:
```diff
@@ -124,7 +144,8 @@ BOOST_AUTO_TEST_CASE(test_PMACTrajectory)
 
   // Try to append once more and verify it fails
-  BOOST_CHECK_EQUAL(trajectory.append(pos, time, user, vel, 10), asynError);
+  // BOOST_CHECK_EQUAL(trajectory.append(pos, time, user, velMode, 10), asynError);
```
Remove commented code

ajgdls commented:
Hi @LeandroMartinsdS mostly minor comments, just the velocity mapping that looks incorrect (V and W mixed).

LeandroMartinsdS committed:
Fix bug in W_Vel in ppmac trajectory scan code

LeandroMartinsdS committed:
Improve trajectory pvt time check

LeandroMartinsdS committed:
Minor fixes: remove whitespaces and commented code

LeandroMartinsdS committed:
Substitute deprecated function in python tests

LeandroMartinsdS committed:
Fix time check for trajectory velocity calculation

LeandroMartinsdS committed:
Remove whitespaces

LeandroMartinsdS commented:
@ajgdls changes are all pushed.
